l3embedding/train.py

```diff
@@ -0,0 +1,420 @@
+
+
+def single_epoch_data_generator(data_dir, epoch_size, **kwargs):
+    while True:
+        data_gen = data_generator(data_dir, **kwargs)
+        for idx, item in enumerate(data_gen):
+            yield item
+            # Once we generate all batches for an epoch, restart the generator
+            if (idx + 1) == epoch_size:
+                break
+
+
+def get_restart_info(history_path):
+    last = None
+    with open(history_path, 'r') as f:
+        reader = csv.DictReader(f)
+        for row in reader:
+            last = row
+
+    return int(last['epoch']), float(last['val_acc']), float(last['val_loss'])
+
+
+def train(train_data_dir, validation_data_dir, output_dir,
+          num_epochs=150, train_epoch_size=512, validation_epoch_size=1024,
+          train_batch_size=64, validation_batch_size=64,
+          model_type='cnn_L3_orig', random_state=20180123,
+          learning_rate=1e-4, verbose=False, checkpoint_interval=10,
+          log_path=None, disable_logging=False, gpus=1, continue_model_dir=None,
+          gsheet_id=None, google_dev_app_name=None):
+
+    init_console_logger(LOGGER, verbose=verbose)
+    if not disable_logging:
+        init_file_logger(LOGGER, log_path=log_path)
+    LOGGER.debug('Initialized logging.')
+
+    # Form model ID
+    data_subset_name = os.path.basename(train_data_dir)
+    data_subset_name = data_subset_name[:data_subset_name.rindex('_')]
+    model_id = os.path.join(data_subset_name, model_type)
+
+    param_dict = {
+          'username': getpass.getuser(),
+          'train_data_dir': train_data_dir,
+          'validation_data_dir': validation_data_dir,
+          'model_id': model_id,
+          'output_dir': output_dir,
+          'num_epochs': num_epochs,
+          'train_epoch_size': train_epoch_size,
+          'validation_epoch_size': validation_epoch_size,
+          'train_batch_size': train_batch_size,
+          'validation_batch_size': validation_batch_size,
+          'model_type': model_type,
+          'random_state': random_state,
+          'learning_rate': learning_rate,
+          'verbose': verbose,
+          'checkpoint_interval': checkpoint_interval,
+          'log_path': log_path,
+          'disable_logging': disable_logging,
+          'gpus': gpus,
+          'continue_model_dir': continue_model_dir,
+          'git_commit': git.Repo(os.path.dirname(os.path.abspath(__file__)),
+                                 search_parent_directories=True).head.object.hexsha,
+          'gsheet_id': gsheet_id,
+          'google_dev_app_name': google_dev_app_name
+    }
+    LOGGER.info('Training with the following arguments: {}'.format(param_dict))
+
+    if continue_model_dir:
+        latest_model_path = os.path.join(continue_model_dir, 'model_latest.h5')
+        m, inputs, outputs = load_model(latest_model_path, model_type, return_io=True, src_num_gpus=gpus)
+    else:
+        m, inputs, outputs = MODELS[model_type](num_gpus=gpus)
+
+    loss = 'binary_crossentropy'
```

Comment:
The loss is the same as as categorical_crossentropy, up to a constant. But binary_crossentropy = 2 * categorical_crossentropy. This shouldn't cause any serious optimization issues, but effectively halves the weight decay constant.